builtin/logical/pki/cert_util.go

```diff
@@ -162,6 +241,291 @@ func validateCommonNames(req *logical.Request, commonNames []string, role *roleE
 	return "", nil
 }
 
+func generateCert(b *backend,
+	role *roleEntry,
+	signingBundle *caInfoBundle,
+	isCA bool,
+	req *logical.Request,
+	data *framework.FieldData) (*certutil.ParsedCertBundle, error) {
+
+	creationBundle, err := generateCreationBundle(b, role, signingBundle, nil, req, data)
+	if err != nil {
+		return nil, err
+	}
+
+	if isCA {
+		creationBundle.IsCA = isCA
+	}
+
+	parsedBundle, err := createCertificate(creationBundle)
+	if err != nil {
+		return nil, err
+	}
+
+	return parsedBundle, nil
+}
+
+// N.B.: This is only meant to be used for generating intermediate CAs.
+// It skips some sanity checks.
+func generateCSR(b *backend,
+	role *roleEntry,
+	signingBundle *caInfoBundle,
+	req *logical.Request,
+	data *framework.FieldData) (*certutil.ParsedCSRBundle, error) {
+
+	creationBundle, err := generateCreationBundle(b, role, signingBundle, nil, req, data)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedBundle, err := createCSR(creationBundle)
+	if err != nil {
+		return nil, err
+	}
+
+	return parsedBundle, nil
+}
+
+func signCert(b *backend,
+	role *roleEntry,
+	signingBundle *caInfoBundle,
+	isCA bool,
+	useCSRValues bool,
+	req *logical.Request,
+	data *framework.FieldData) (*certutil.ParsedCertBundle, error) {
+
+	csrString := data.Get("csr").(string)
+	if csrString == "" {
+		return nil, certutil.UserError{Err: fmt.Sprintf("\"csr\" is empty")}
+	}
+
+	pemBytes := []byte(csrString)
+	pemBlock, pemBytes := pem.Decode(pemBytes)
+	if pemBlock == nil {
+		return nil, certutil.UserError{Err: "csr contains no data"}
+	}
+	csr, err := x509.ParseCertificateRequest(pemBlock.Bytes)
+	if err != nil {
+		return nil, certutil.UserError{Err: "certificate request could not be parsed"}
+	}
+
+	creationBundle, err := generateCreationBundle(b, role, signingBundle, csr, req, data)
+	if err != nil {
+		return nil, err
+	}
+
+	if isCA {
+		creationBundle.IsCA = isCA
+		creationBundle.UseCSRValues = useCSRValues
```

Comment:
I'm deferring this for now. If there is enough clamoring from users to allow an endpoint that uses CSR values verbatim for *all* certificates we can add that in, but the whole point is to use role-based controls, and it becomes extremely difficult to reason about the final state of a cert if CSR values are used, which may include extensions that this backend doesn't know about, but you're trying to apply a role-based approach. In any reasonably common usage of this backend I don't see this being a problem, so let's not support that until/unless there's demand for it.